cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java

```diff
@@ -515,6 +518,81 @@ public void setInstances(ProgramId programId, int instances, @Nullable String co
     }
   }
 
+  /**
+   * Lists all dataset instances in a namespace. If perimeter security and authorization are enabled, only returns the
+   * dataset instances that the current user has access to.
+   *
+   * @param namespaceId the namespace to list datasets for
+   * @return the dataset instances in the provided namespace
+   */
+  public List<ProgramRecord> list(String namespaceId, ProgramType type) throws Exception {
```

Comment:
can we just accept ``NamespaceId`` in this method, and remove the private method on line 534? This method doesn't add too much value.